Support freeform text notes on uploads (896)

NatHillardUSDR opened the pull request:
### Ticket #896 
## Description
Adds a form to set reportingId, agencyId, and notes in addition to uploading a file, using Vue-Formulate

## Screenshots / Demo Video
![Screenshot 2023-03-04 at 4 33 20 PM](https://user-images.githubusercontent.com/120750336/222929635-692da637-5681-4846-bb16-ad94bcfa6865.jpg)
![Screenshot 2023-03-04 at 4 35 40 PM](https://user-images.githubusercontent.com/120750336/222929662-7996d4c2-3c07-4dd8-8253-c315fb53eaf6.jpg)


## Testing
NOTE! This patch needs to be applied alongside https://github.com/usdigitalresponse/usdr-gost/pull/1035 for the functionality to work! 
Once that has been merged / this is made on top, you should be able to go to Uploads -> Submit Workbook and specify an agency and reporting period

### Automated and Unit Tests
- [ ] Added Unit tests 
(N/A - we do not yet have FE unit tests)

### Manual tests for Reviewer
- [x] Added steps to test feature/functionality manually

## Checklist
- [x] Provided ticket and description
- [x] Provided screenshots/demo
- [x] Provided testing information
- [x] Provided adequate test coverage for all new code
- [x] Added PR reviewers

NOTE: Setting as DRAFT for now in case @ajhyndman has another patch in the works / is planning on a different fix! Currently coordinating in slack.

ajhyndman commented on packages/client/src/arpa_reporter/views/NewUpload.vue:
```diff
@@ -1,111 +1,164 @@
+        },
+        {
+          component: 'div',
+          class: 'row',
```
I noticed there's a slightly wonky layout effect here.  I think the issue is that the `row` class isn't meant to be used without a corresponding `col`.

Probably the easiest fix is just to swap this for `d-flex`!
```suggestion
          class: 'd-flex',
```

NatHillardUSDR replied:
Ah, good catch!

ajhyndman commented on packages/client/src/arpa_reporter/views/NewUpload.vue:
```diff
@@ -1,111 +1,164 @@
+    outer: 'form-group',
+    input: (context, classes) => {
+      // set this to 'form-control' unless it's a file input
+      const inputClass = context.type === 'file' ? '' : 'form-control';
```
```suggestion
      const inputClass = context.type === 'file' ? 'form-control-file' : 'form-control';
```

There is actually a form-control-file class we can use here.  It looks like it doesn't change anything visually, but it does apply `display: block` to the element, creating a larger click target, which seems nice.

NatHillardUSDR replied:
Ah cool - thanks for catching!

ajhyndman commented on packages/client/src/arpa_reporter/views/NewUpload.vue:
```diff
@@ -1,111 +1,164 @@
+    input: (context, classes) => {
+      // set this to 'form-control' unless it's a file input
+      const inputClass = context.type === 'file' ? '' : 'form-control';
+      return [classes, inputClass].join(' ');
```
According to the documentation, if we return an array here, it will automatically join the new input class with existing classes.

```suggestion
      return [inputClass];
```

ajhyndman commented on packages/client/src/arpa_reporter/views/NewUpload.vue:
```diff
@@ -1,111 +1,164 @@
 <template>
-  <div class="upload">
-    <h1>Submit Workbook</h1>
```
I'd prefer to bring back the document heading here!

NatHillardUSDR replied:
🤦

ajhyndman commented on packages/client/src/arpa_reporter/views/NewUpload.vue:
```diff
@@ -1,111 +1,164 @@
 
+Vue.use(VueFormulate, {
+  classes: {
+    prefix: 'formulate-',
```
What is this prefix class for?  It doesn't obviously change anything in my testing.

NatHillardUSDR replied:
Ah, interesting - this is actually doubly wrong. 

I believe I had seen a reference to this in an issue in the repo, as a means of prefixing class names so as to avoid conflicts with other library class names. The property is apparently actually named `idPrefix`, and is a sibling to `classes`. 

Also though, by [default](https://github.com/wearebraid/vue-formulate/blob/3fdd6fdc2ed82113e6a78dfeec800b36763d7f95/src/Formulate.js#L96) it's `formulate-`, so i can just remove this line.

ajhyndman commented on packages/client/src/arpa_reporter/views/NewUpload.vue:
```diff
@@ -1,111 +1,164 @@
-      files: null,
-      uploading: false,
+      formErrors: [],
+      initialValues: () => ({
+        reportingPeriodId: this.$store.getters.viewPeriodID ?? '',
+        agencyId: this.$store.getters.user.agency.id,
+      }),
```
Could we move this into `methods` alongside `reportingPeriodSelectItems` and `agencySelectItems`?

NatHillardUSDR replied:
👍 I had initially moved it here due to the issue I go into below, but if I move to methods and then set properly it works

ajhyndman commented on packages/client/src/arpa_reporter/views/NewUpload.vue:
```diff
@@ -1,111 +1,164 @@
+        agencyId: this.$store.getters.user.agency.id,
+      }),
+
+      values: {},
```
```suggestion
      values: this.initialValues(),
```

Instead of populating this in the `created` lifecycle hook, can we just populate it here?

NatHillardUSDR replied:
I had briefly tried this before only to find it didn't populate correctly. That said! Trying it again, I'm realizing now, in having used an arrow function, I hadn't sufficiently captured `this`, which was the true culprit!

ajhyndman replied:
Ahh, nice catch!  Yeah, that would be a sneaky one.

ajhyndman commented on packages/client/src/arpa_reporter/views/NewUpload.vue:
```diff
@@ -1,111 +1,164 @@
+         we just return the file object. */
+      return file;
+    },
+    async onSubmit(formData) {
```
```suggestion
    async onSubmit(data) {
```

Nit: just for clarity, could we give this a name that doesn't imply it is an instance of the `FormData` class?

NatHillardUSDR replied:
A good call, fixed

ajhyndman commented on packages/client/src/arpa_reporter/views/NewUpload.vue:
```diff
@@ -1,111 +1,164 @@
+      return file;
+    },
+    async onSubmit(formData) {
+      // This is technically a FileList, allowing for multiple uploads if configured
+      // this way, but for now we have only a single file.
+      const file = formData.spreadsheet[0];
+      const uploadData = {
+        reportingPeriodId: formData.reportingPeriodId,
+        agencyId: formData.agencyId,
+        expenditure_code: formData.expenditure_code,
+        notes: formData.notes,
+      };
+      const uploadFormData = new FormData();
+      uploadFormData.append('spreadsheet', file);
+      Object.keys(uploadData).forEach((key) => {
+        const value = uploadData[key];
+        if (value) {
+          uploadFormData.append(key, value);
+        }
+      });
```
```suggestion
      const uploadFormData = new FormData();
      uploadFormData.append('spreadsheet', data.spreadsheet[0]);
      Object.entries(_.omit(data, ['spreadsheet'])).forEach(([key, value]) => {
        if (value) {
          uploadFormData.append(key, value);
        }
      });
```

Nit: We do have lodash available in this project.  Using that, we can probably simplify this block a little!

NatHillardUSDR replied:
Ooh, I have less familiarity with all the functionality available in lodash - nice to have this!

ajhyndman replied:
I have found the lodash docs to be pretty good!  I do still have to check them pretty frequently to double-check the exact syntax.

https://lodash.com/docs/

ajhyndman commented on packages/client/src/main.js:
```diff
@@ -15,6 +15,7 @@ Vue.use(BootstrapVue);
 // Optionally install the BootstrapVue icon components plugin
 Vue.use(IconsPlugin);
 Vue.use(Vuelidate);
+Vue.use(VueFormulate);
```
I think it isn't necessary to mount VueFormulate globally, is it?

NatHillardUSDR replied:
I tried various variations of this - It seems to only work with mounting it on a per-file basis. It also doesn't play well with exports/imports, or with attempts to register it under `components`... 

Removing it globally and keeping it per-file for now seems to work, I can keep it this way.

ajhyndman reviewed (changes requested):
Overall, I'm quite happy with this approach!

I went through and left some polish suggestions, but they are all pretty minor.

NatHillardUSDR commented:
Thanks @ajhyndman for your review! I appreciate your thoughtful approach and suggestions, and I learned a few things in the process. 

I made your recommended fixes, made a few additional, then rebased to clean up earlier commits / rebased against `_staging` as well. Taking out of draft, as #1035 just landed, which I also rebased against

TylerHendrickson reviewed (approved):
Looks great!

NatHillardUSDR commented:
Just now I:
1. Fixed a few linting errors on files that were noted by the CI job, and
2. Updated the reportingPeriodId dropdown to be readonly, per the AC spelled out in #786 

Here's a screenshot with the field disabled - note it still sends the value in the BE:
![Screenshot 2023-03-06 at 1 26 49 PM](https://user-images.githubusercontent.com/120750336/223200820-2da2accb-f2f9-4eb4-8fc6-92ca97d5cfd5.jpg)

Thank you @TylerHendrickson for the review; @ajhyndman can you take a look as well to see if this addresses the issues you called out?

ajhyndman commented on packages/client/src/arpa_reporter/views/NewUpload.vue:
```diff
@@ -1,113 +1,154 @@
+      :schema="schema"
+      @submit="onSubmit"
+    >
+    <FormulateErrors
+      class="alert alert-danger"
+    />
```
```suggestion
      <FormulateErrors
        class="alert alert-danger"
      />
```

Nit: The indentation looks slightly off to me here.

ajhyndman commented on packages/client/src/main.js:
```diff
@@ -2,7 +2,6 @@ import Vue from 'vue';
 import { BootstrapVue, IconsPlugin } from 'bootstrap-vue';
 import vSelect from 'vue-select';
 import Vuelidate from 'vuelidate';
-
```
Nit: I think we lost this blank line during a revert.

ajhyndman reviewed (approved):
Thanks!  This looks great.

I left two formatting nits, but honestly you should probably just ignore them for now.  We should set up an auto-formatter (like Prettier) at some point.  That would fix this too.

NatHillardUSDR commented:
Thanks! The formatting issues you call out are odd, I'm not seeing these on my branch - it's possible this arose due to having rebased + force pushed? I can look to avoid this workflow in the future actually, as if you've pulled locally weird things can happen.

NatHillardUSDR commented on packages/client/src/arpa_reporter/views/NewUpload.vue:
```diff
@@ -1,113 +1,154 @@
+        {
+          type: 'text',
+          label: 'Expenditure Code',
+          name: 'expenditure_code',
```
Ah interesting - I had gotten this from this screenshot: 
![image](https://user-images.githubusercontent.com/120750336/224113359-b0e4c197-44f5-41d5-b457-e63b5177c508.png)
in the original supplemental data parent [issue](https://github.com/usdigitalresponse/usdr-gost/issues/786).

Though actually, it's true, this is not used by `createUploadRow` which actually determines the data inserted into the DB. I can remove this for now.
